### PR TITLE
[DPE-3684][DPE-6020] Reinitialise raft

### PR DESCRIPTION
**Pull request by dragomirp**
Syncobj RAFT implementation, used as a standalone DCS for Patroni, cannot elect a leader if the cluster loses quorum and becomes read only. This will prevent Patroni from automatically switching over, even in cases where sync_standbys are available in the cluster and could take over as primary.

The PR adds logic to detect when the RAFT cluster becomes read only and updates `promote-to-primary` action to be able to trigger a RAFT re-initialisation.

Contains changes from https://github.com/canonical/postgresql-operator/pull/663

### CODE REVIEW

File: tests/integration/ha_tests/test_scaling.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    start_continuous_writes,
+)
+
+logger = logging.getLogger(__name__)
+
+charm = None
+
+
+@pytest.mark.group(1)
+@markers.juju3
```
Comment (dragomirp) [inline tests/integration/ha_tests/test_scaling.py:31]: Only tested on juju 3 for the moment. Should be able to reuse the wrapper around force removal to enable the tests on juju 2

File: src/cluster.py [hunk 2]
```diff
@@ -721,32 +793,59 @@ def remove_raft_member(self, member_ip: str) -> None:
...
+                health_status = self.get_patroni_health()
+            except Exception:
+                logger.warning("Remove raft member: Unable to get health status")
+                health_status = {}
+            if health_status.get("role") in ("leader", "master") or health_status.get(
+                "sync_standby"
+            ):
+                logger.info("%s is raft candidate" % self.charm.unit.name)
+                data_flags["raft_candidate"] = "True"
+            self.charm.unit_peer_data.update(data_flags)
```
Comment (dragomirp) [inline src/cluster.py:830]: Initial flags that trigger the recovery attempt. If we need to disable this (stable releases?), it would be easiest from here.

File: src/cluster.py [hunk 3]
```diff
@@ -710,6 +731,57 @@ def primary_changed(self, old_primary: str) -> bool:
...
+    def reinitialise_raft_data(self) -> None:
+        """Reinitialise the raft journals and promoting the unit to leader. Should only be run on sync replicas."""
+        logger.info("Rerendering patroni config without peers")
+        self.charm.update_config(no_peers=True)
+        logger.info("Starting patroni")
+        self.start_patroni()
+
+        logger.info("Waiting for new raft cluster to initialise")
+        for attempt in Retrying(wait=wait_fixed(5)):
+            with attempt:
```
Comment (dragomirp) [inline src/cluster.py:764]: We have to define sane timeout for the cases when the recovery hijacks execution.

File: src/cluster.py [hunk 4]
```diff
@@ -585,8 +604,10 @@ def render_patroni_yml_file(
             data_path=POSTGRESQL_DATA_PATH,
             enable_tls=enable_tls,
             member_name=self.member_name,
-            partner_addrs=self.charm.async_replication.get_partner_addresses(),
-            peers_ips=self.peers_ips,
+            partner_addrs=self.charm.async_replication.get_partner_addresses()
+            if not no_peers
+            else [],
+            peers_ips=self.peers_ips if not no_peers else set(),
```
Comment (dragomirp) [inline src/cluster.py:610]: Ignoring the other nodes when recovering, they should rejoin afterwards

File: src/charm.py [hunk 5]
```diff
@@ -508,20 +508,163 @@ def _on_pgdata_storage_detaching(self, _) -> None:
...
+                if flag.startswith("raft_"):
+                    return
+
+        logger.info("Cleaning up raft app data")
+        self.app_peer_data.pop("raft_rejoin", None)
+        self.app_peer_data.pop("raft_reset_primary", None)
+        self.app_peer_data.pop("raft_selected_candidate", None)
+        self.app_peer_data.pop("raft_followers_stopped", None)
+
+    def _raft_reinitialisation(self) -> None:
```
Comment (dragomirp) [inline src/charm.py:591]: If there's only one unit (sync standby and leader), this should execute in one go.

File: src/charm.py [hunk 6]
```diff
@@ -508,20 +508,163 @@ def _on_pgdata_storage_detaching(self, _) -> None:
...
             logger.debug("Deferring on_peer_relation_changed: cluster not initialized")
             event.defer()
-            return
+            return False
+
+        # Check whether raft is stuck.
+        if self.has_raft_keys():
+            self._raft_reinitialisation()
+            logger.debug("Early exit on_peer_relation_changed: stuck raft recovery")
+            return False
```
Comment (dragomirp) [inline src/charm.py:654]: This will hijack execution until recovery completes. We should think of a ways to detect manual recovery.

File: src/cluster.py [hunk 7]
```diff
@@ -721,32 +793,59 @@ def remove_raft_member(self, member_ip: str) -> None:
...
+            not raft_status["leader"] or raft_status["leader"].host == member_ip
+        ):
+            logger.warning("Remove raft member: Stuck raft cluster detected")
+            data_flags = {"raft_stuck": "True"}
+            try:
+                health_status = self.get_patroni_health()
+            except Exception:
+                logger.warning("Remove raft member: Unable to get health status")
+                health_status = {}
+            if health_status.get("role") in ("leader", "master") or health_status.get(
```
Comment (dragomirp) [inline src/cluster.py:825]: We shouldn't have a stuck cluster with leader role, but just in case.

File: src/charm.py [hunk 8]
```diff
@@ -508,20 +508,163 @@ def _on_pgdata_storage_detaching(self, _) -> None:
...
+        if not raft_stuck:
+            return
+
+        if not all_units_stuck:
+            logger.warning("Stuck raft not yet detected on all units")
+            return
+
+        if not candidate:
+            logger.warning("Stuck raft has no candidate")
+            return
```
Comment (dragomirp) [inline src/charm.py:535]: We can't proceed with automatic recovery if there's no sync standby, since the first unit on the new raft cluster will be leader and there may be data loss if promoting an async replica. We should consider setting a status and ways for manual recovery, if the user wants to promote a given replica anyway.

File: src/cluster.py [hunk 9]
```diff
@@ -721,32 +793,59 @@ def remove_raft_member(self, member_ip: str) -> None:
...
         # Check whether the member is still part of the raft cluster.
-        if not member_ip or member_ip not in raft_status:
+        if not member_ip or f"partner_node_status_server_{member_ip}:2222" not in raft_status:
+            return
+
+        # If there's no quorum and the leader left raft cluster is stuck
+        if not raft_status["has_quorum"] and (
+            not raft_status["leader"] or raft_status["leader"].host == member_ip
+        ):
+            logger.warning("Remove raft member: Stuck raft cluster detected")
```
Comment (dragomirp) [inline src/cluster.py:818]: We should most probably set a status here to indicate what's going on.

File: src/cluster.py [hunk 10]
```diff
@@ -710,6 +731,57 @@ def primary_changed(self, old_primary: str) -> bool:
         primary = self.get_primary()
         return primary != old_primary
 
+    def remove_raft_data(self) -> None:
```
Comment (dragomirp) [inline src/cluster.py:734]: We can most probably converge this with https://github.com/canonical/postgresql-operator/blob/main/src/relations/async_replication.py#L695 as a followup

**Review by delgod — APPROVED** [review]

**Review by lucasgameiroborges — APPROVED** [review]
LGTM! I just have one theoretical question:

Suppose a cluster with 5 nodes: 1 leader, 2 sync replicas and 2 async replicas. Lets say for some reason that this cluster loses connection in a weird way and now it got split in 2 partitions: leader + sync + async (partition 1), sync + async (partition 2). 

Since we eliminated the need for a quorum when electing a new leader, couldn't the sync node in partition 2 elect itself as new leader, partition 1 keeps working the same way bc of quorum, and now we have 2 distinct clusters?

It's a weird edge case yes, just curious if we have any answer to it.

File: tests/integration/ha_tests/test_scaling.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    logger.info("Deleting sync replica")
+    primary = await get_primary(ops_test, app)
+    secondary = next(
+        filter(lambda x: x.name != primary, ops_test.model.applications[DATABASE_APP_NAME].units)
+    ).name
+    await ops_test.model.destroy_unit(secondary, force=True, destroy_storage=False, max_wait=1500)
+
+    await ops_test.model.wait_for_idle(status="active", timeout=600, idle_period=45)
+
+    await are_writes_increasing(ops_test, secondary)
```
Comment (lucasgameiroborges) [inline tests/integration/ha_tests/test_scaling.py:109]: Another theoretical question here: In a 2-node cluster, primary + sync replica, if the sync replica goes down/gets unresponsive, how can the primary continue to accept write requests (writes increasing)? Isn't the whole point of having a sync replica to make sure that every write operation gets recognized by another node besides the primary?

Comment (dragomirp) [reply tests/integration/ha_tests/test_scaling.py:109]: > Since we eliminated the need for a quorum when electing a new leader, couldn't the sync node in partition 2 elect itself as new leader, partition 1 keeps working the same way bc of quorum, and now we have 2 distinct clusters?

We rely on Juju to know the full number of units. If not all units detect a loss of quorum, we don't reinitialise the raft cluster. The Juju leader selects the new primary when looping over the potential candidates and puts it in the app peer data.

> Another theoretical question here: In a 2-node cluster, primary + sync replica, if the sync replica goes down/gets unresponsive, how can the primary continue to accept write requests (writes increasing)? Isn't the whole point of having a sync replica to make sure that every write operation gets recognized by another node besides the primary?

There's logic to maintain a list of units in the peer data, that should detect that the unit is gone and downgrade the cluster to single node.

Comment (lucasgameiroborges) [reply tests/integration/ha_tests/test_scaling.py:109]: that makes sense, thanks :)

**Review by taurus-forever — APPROVED** [review]
LGTM after a lot of testing and edge cases discussions. THANK YOU!

**Review by marceloneppel — APPROVED** [review]
LGTM! Thanks, Dragomir!

File: src/charm.py [hunk 12]
```diff
@@ -508,25 +518,171 @@ def _on_pgdata_storage_detaching(self, _) -> None:
...
+
+            if self.unit.is_leader():
+                self._stuck_raft_cluster_rejoin()
+
+        if "raft_rejoin" in self.app_peer_data:
+            logger.info("Cleaning up raft unit data")
+            self.unit_peer_data.pop("raft_primary", None)
+            self.unit_peer_data.pop("raft_stopped", None)
+            self.update_config()
+            self._patroni.start_patroni()
```
Comment (dragomirp) [inline src/charm.py:639]: Restarting the non-primary units.

File: src/charm.py [hunk 13]
```diff
@@ -508,25 +518,171 @@ def _on_pgdata_storage_detaching(self, _) -> None:
...
+                continue
+            if "raft_primary" in data:
+                primary = key
+                break
+        if primary and "raft_reset_primary" not in self.app_peer_data:
+            logger.info("Updating the primary endpoint")
+            self.app_peer_data.pop("members_ips", None)
+            for unit in self._peers.units:
+                self._add_to_members_ips(self._get_unit_ip(unit))
+            self._add_to_members_ips(self._get_unit_ip(self.unit))
```
Comment (dragomirp) [inline src/charm.py:564]: If the primary and leader are different, cluster will be unable to reconfigure, since the leader patroni is down and outside the cluster, so we have to keep the list here.

**Review by marceloneppel — APPROVED** [review]

**Comment by nobuto-m** [comment]
@taurus-forever gave me a charm branch (`14/edge/pr611`) of testing this in the context of the following issue.
https://github.com/canonical/postgresql-operator/issues/571

However, it's not working for me. Could you please take a look into my steps if I'm doing something wrong or a patch could be incomplete?

[postgresql_debug.log](https://github.com/user-attachments/files/17504382/postgresql_debug.log)
```
unit-postgresql-2: 08:25:06 INFO juju.worker.uniter found queued "leader-elected" hook
unit-postgresql-2: 08:25:10 WARNING unit.postgresql/2.juju-log Failed to connect to PostgreSQL.
unit-postgresql-2: 08:25:15 WARNING unit.postgresql/2.juju-log Failed to connect to PostgreSQL.
unit-postgresql-2: 08:25:20 WARNING unit.postgresql/2.juju-log Failed to connect to PostgreSQL.
unit-postgresql-2: 08:25:26 WARNING unit.postgresql/2.juju-log Failed to connect to PostgreSQL.
unit-postgresql-2: 08:25:31 WARNING unit.postgresql/2.juju-log Failed to connect to PostgreSQL.
unit-postgresql-2: 08:25:31 INFO juju.worker.uniter.operation ran "leader-elected" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:27:05 WARNING unit.postgresql/2.juju-log database-peers:0: Remove raft member: Stuck raft cluster detected
unit-postgresql-2: 08:27:05 WARNING unit.postgresql/2.juju-log database-peers:0: Stuck raft not yet detected on all units
unit-postgresql-2: 08:27:10 WARNING unit.postgresql/2.juju-log database-peers:0: Failed to connect to PostgreSQL.
unit-postgresql-2: 08:27:15 WARNING unit.postgresql/2.juju-log database-peers:0: Failed to connect to PostgreSQL.
unit-postgresql-2: 08:27:20 WARNING unit.postgresql/2.juju-log database-peers:0: Failed to connect to PostgreSQL.
unit-postgresql-2: 08:27:25 WARNING unit.postgresql/2.juju-log database-peers:0: Failed to connect to PostgreSQL.
unit-postgresql-2: 08:27:30 WARNING unit.postgresql/2.juju-log database-peers:0: Failed to connect to PostgreSQL.
unit-postgresql-2: 08:27:30 INFO juju.worker.uniter.operation ran "database-peers-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:27:31 INFO juju.worker.uniter.operation ran "restart-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:27:32 INFO juju.worker.uniter.operation ran "upgrade-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:27:33 WARNING unit.postgresql/2.juju-log database-peers:0: Stuck raft not yet detected on all units
unit-postgresql-2: 08:27:33 INFO juju.worker.uniter.operation ran "database-peers-relation-changed" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:27:49 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:32:12 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:36:35 INFO juju.worker.uniter.operation ran "database-peers-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:36:36 INFO juju.worker.uniter.operation ran "restart-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:36:37 INFO juju.worker.uniter.operation ran "upgrade-relation-departed" hook (via hook dispatching script: dispatch)
unit-postgresql-2: 08:36:49 INFO juju.worker.uniter.operation ran "update-status" hook (via hook dispatching script: dispatch)
```

### Steps

1. juju deploy postgresql -n 3 --channel 14/edge/pr611 --base ubuntu@22.04
2. check the topology

```
$ sudo -u snap_daemon patronictl -c /var/snap/charmed-postgresql/current/etc/patroni/patroni.yaml topology
+ Cluster: postgresql (7429252738803380683) -+-----------+----+-----------+
| Member         | Host       | Role         | State     | TL | Lag in MB |
+----------------+------------+--------------+-----------+----+-----------+
| postgresql-1   | 10.0.9.118 | Leader       | running   |  1 |           |
| + postgresql-0 | 10.0.9.67  | Sync Standby | streaming |  1 |         0 |
| + postgresql-2 | 10.0.9.77  | Replica      | streaming |  1 |         0 |
+----------------+------------+--------------+-----------+----+-----------+
```

3. take down two nodes including the leader

```
lxc stop -f juju-c3fd60-1
lxc stop -f juju-c3fd60-0
```

4. check the status

```
$ juju status
Model        Controller  Cloud/Region         Version  SLA          Timestamp
postgres-pr  localhost   localhost/localhost  3.5.4    unsupported  08:24:32Z

App         Version  Status  Scale  Charm       Channel        Rev  Exposed  Message
postgresql  14.13    active    1/3  postgresql  14/edge/pr611  504  no       

Unit           Workload  Agent  Machine  Public address  Ports     Message
postgresql/0   unknown   lost   0        10.0.9.67       5432/tcp  agent lost, see 'juju show-status-log postgresql/0'
postgresql/1*  unknown   lost   1        10.0.9.118      5432/tcp  agent lost, see 'juju show-status-log postgresql/1'
postgresql/2   active    idle   2        10.0.9.77       5432/tcp  

Machine  State    Address     Inst id        Base          AZ  Message
0        down     10.0.9.67   juju-c3fd60-0  ubuntu@22.04      Running
1        down     10.0.9.118  juju-c3fd60-1  ubuntu@22.04      Running
2        started  10.0.9.77   juju-c3fd60-2  ubuntu@22.04      Running
```

```
$ sudo -u snap_daemon env PATRONI_LOG_LEVEL=DEBUG patronictl -c /var/snap/charmed-postgresql/current/etc/patroni/patroni.yaml topology
2024-10-24 08:26:12,698 - DEBUG - Loading configuration from file /var/snap/charmed-postgresql/current/etc/patroni/patroni.yaml
2024-10-24 08:26:17,946 - INFO - waiting on raft
2024-10-24 08:26:22,947 - INFO - waiting on raft
2024-10-24 08:26:27,948 - INFO - waiting on raft
^C
Aborted!
```

Both are expected.


5. Remove units one by one

```
$ juju remove-unit postgresql/0 --force
WARNING This command will perform the following actions:
will remove unit postgresql/0
- will remove storage pgdata/0

Continue [y/N]? y
```
```
$ juju remove-unit postgresql/1 --force
WARNING This command will perform the following actions:
will remove unit postgresql/1
- will remove storage pgdata/1

Continue [y/N]? y
```

6. check the status once again

```
$ juju status
Model        Controller  Cloud/Region         Version  SLA          Timestamp
postgres-pr  localhost   localhost/localhost  3.5.4    unsupported  08:45:42Z

App         Version  Status  Scale  Charm       Channel        Rev  Exposed  Message
postgresql  14.13    active      1  postgresql  14/edge/pr611  504  no       

Unit           Workload  Agent  Machine  Public address  Ports     Message
postgresql/2*  active    idle   2        10.0.9.77       5432/tcp  

Machine  State    Address    Inst id        Base          AZ  Message
2        started  10.0.9.77  juju-c3fd60-2  ubuntu@22.04      Running
```

```
$ juju run postgresql/leader get-primary
Running operation 3 with 1 task
  - task 4 on unit-postgresql-2

Waiting for task 4...
primary: postgresql/1
```

^^^ This should return `postgresql/2` since that's the only member remaining in the model.

And the raft membership has a problem.

```
$ sudo -u snap_daemon env PATRONI_LOG_LEVEL=DEBUG patronictl -c /var/snap/charmed-postgresql/current/etc/patroni/patroni.yaml topology
2024-10-24 08:39:04,529 - DEBUG - Loading configuration from file /var/snap/charmed-postgresql/current/etc/patroni/patroni.yaml
2024-10-24 08:39:09,759 - INFO - waiting on raft
2024-10-24 08:39:14,759 - INFO - waiting on raft
2024-10-24 08:39:19,760 - INFO - waiting on raft
^C
Aborted!
```

**Comment by taurus-forever** [comment]
Dear @nobuto-m , thank you for testing and sharing us feedback!

Please check this the PR description: `The PR adds logic to detect when the RAFT cluster becomes read only and to reinitialise it, **if a sync_standby is available** to become a primary.`

The reported test case has stopped both Leader and Sync_Standby simultaneously.
Automated promotion in this case is not triggered by design,
as we cannot guarantee consistency for Replica type of the node.

We discussed this case on the last sync call: 
 * (recommended) user restores one of the offline node (raft unlocks and old Primary/Sync_Standby will be elected as a new Primary)
 * Patroni/Raft cluster should be re-initialised (manually now). We will create the new force-something action (separate PR). It should be executed as a last resort if user accepts risks of loosing the last transaction(s).
 * in three nodes setup always keep: Primary + 2*Sync_Standby => noticeable write performance impact.
No other options on the table.

We are going to improve UX and Juju statuses to inform users better about the Replica only left in the cluster.
Do you see other improvements/fixes necessary here? Thank you!

**Comment by taurus-forever** [comment]
```
L # Primary  Sync-Standby    Replica
* 0                                   # All OK: RW+RO
* 1                             X     # Degraded mode RW+RO: no changes.
  2               X                   # Degraded mode RW+RO: Replica -> Sync-Standby
  3   X                               # Degraded mode RW+RO: Sync-Standby -> Primary, Replica -> Sync-Standby
* 4               X             X     # Disaster mode: re-add/restore units ASAP
  5   X                         X     # Disaster mode: Sync-Standby -> Primary AFTER juju remove-units (PR #611)
  6   X           X                   # Disaster mode: requires manual confirmation of Replica -> Primary. NEW action.
  7   X           X             X     # Disaster mode: reinstall and restore from backup or reuse Primary/SyncStandby disk
```

this PR addresses case #5, can you please check it from your side.

**Comment by taurus-forever** [comment]
Based on the discussion I have re-checked the cluster installation speed and the case 4:

* The fresh 3-nodes cluster installation on my laptop takes 00:05:12.
The snap downloading time is close to 1 minute on each unit.

* Re-tested the case #4, the raft issues detected after `juju remove-unit --force`:
```
unit-postgresql-2: 19:07:29 WARNING unit.postgresql/2.juju-log database-peers:0: Stuck raft has no candidate                         
unit-postgresql-2: 19:07:29 DEBUG unit.postgresql/2.juju-log database-peers:0: Early exit on_peer_relation_changed: stuck raft recovery  
```
but for some reason the raft re-initialization didn't happen in full.

@dragomirp can you please check the case 4 in https://github.com/canonical/postgresql-operator/pull/611#issuecomment-2435693133 and make sure the same logic applied here as in case 5. Tnx!

* as for the case 6: we will approve the dedicated juju action name and prepare the separate PR/commit in the same GH issue.
